Doc/reference/executionmodel.rst

```diff
@@ -163,14 +166,65 @@ These references follow the normal rules for name resolution with an exception
 that unbound local variables are looked up in the global namespace.
 The namespace of the class definition becomes the attribute dictionary of
 the class. The scope of names defined in a class block is limited to the
-class block; it does not extend to the code blocks of methods -- this includes
+class block; it does not extend to the code blocks of methods. This includes
 comprehensions and generator expressions since they are implemented using a
-function scope.  This means that the following will fail::
+function scope, but it does not include :ref:`annotation scopes <annotation-scopes>`,
+which have access to their enclosing class scopes.
+This means that the following will fail::
 
    class A:
        a = 42
        b = list(a + i for i in range(10))
 
+However, the following will succeed::
+
+   class A:
+       type Alias = Nested
+       class Nested: ...
+
+   print(A.Alias.__value__)  # <type 'A.Nested'>
+
+.. _annotation-scopes:
+
+Annotation scopes
+-----------------
+
+:ref:`Type parameter lists <type-params>` and :keyword:`type` statements
+introduce *annotation scopes*, which behave mostly like function scopes,
+but with some exceptions discussed below. :term:`Annotations <annotation>`
+currently do not use annotation scopes, but they are expected to use
+annotation scopes in Python 3.13 when :pep:`649` is implemented.
+
+Annotation scopes are used in the following contexts:
+
+* Type parameter lists for generic type aliases.
+* Type parameter lists for generic functions. The function's annotations are
+  executed within the type scope, but its defaults and decorators are not.
+* Type parameter lists for generic classes. The class's base classes and
+  keyword arguments are executed within the type scope, but its decorators are not.
+* The bounds and constraints for type variables.
+* The value of type aliases.
+
+Annotation scopes differ from function scopes in the following ways:
+
+* If an annotation scope is immediately within a class scope, or within another
+  annotation scope that is immediately within a class scope, names defined in the
+  class scope can be accessed from within the annotation scope, as if the code in the
+  annotation scope was executing directly within the class scope. (By contrast, regular
+  functions defined within classes cannot access names defined in the class scope.)
+* Expressions in annotation scopes cannot contain :keyword:`yield`, ``yield from``,
+  :keyword:`await`, or :token:`:= <~python-grammar:expression>` expressions. (These
+  expressions are allowed in other scopes contained within the type scope.)
```

Comment:
Should we avoid introducing "type scope" as a synonym for "annotation scope"?
```suggestion
  expressions are allowed in other scopes contained within the annotation scope.)
```

Reply:
I was going to use the term "type scope" at first but didn't catch all places where I had used the term, thanks for catching that. (There are a few more in this file.)